### PR TITLE
feat(Core/Creature): Implement quest_greetings table

### PR DESCRIPTION
**Pull request by Kitzunu**
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/a64d1ec51eb11fd01a26d08a09de3d8d7dd89675)

* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/6dda09818d32fc715549cda3e8bd6fa806425b27)

Co-Authored-By: ForesterDev <11771800+ForesterDev@users.noreply.github.com>
Co-Authored-By: Giacomo Pozzoni <giacomopoz@gmail.com>

## Changes Proposed:
-  Implement quest_greetings table
-  https://github.com/azerothcore/wiki/pull/686

## Issues Addressed:

- Closes 

## Tests Performed:

- Builds
- Tested in-game

image

## How to Test the Changes:

1. 
```sql
INSERT INTO `quest_greeting` (`ID`, `Type`, `GreetEmoteType`, `GreetEmoteDelay`, `Greeting`, `VerifiedBuild`) VALUES (341, 0, 0, 0, 'I don\'t have much time for idle talk, $N.  I\'ve got to get this bridge rebuilt before the rains come.  I\'ve finished every project on-time and under budget and I\'m not about to start slipping now.', 0);
```

2. .go c id 341
3. Open gossip and see his quest_greeting is showing if there are 2 or more options there

to get it to show:

.q add 345

Also check that other gossips and quest givers works as they should

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Alastro-Kraken** [comment]
This kind of gossip text is only supposed to be shown if there are more than one gossip options (including quests) the player can choose from, for example one quest and one gossip option, or two quests. Otherwise it should just skip to the quest text.

**Comment by Kitzunu** [comment]
It should not just skip to the quest text, for some creatures. Which is what this PR fixes.
And these types of texts are not "gossips" per se either, as you cannot find any related gossip menu id's for them in sniffs

But I do not understand what you mean by "So the linked image is wrong behavior" can you please explain?

**Comment by neifion-00000000** [comment]
installed, tested, and working

image

**Comment by neifion-00000000** [comment]
> 

These come from SMSG_QUEST_GIVER_QUEST_LIST_MESSAGE and are correct/blizzlike 

this is not the same as gossip, which is dealt with separately

**Comment by neifion-00000000** [comment]
> This kind of gossip text is only supposed to be shown if there are more than one gossip options (including quests) the player can choose from, for example one quest and one gossip option, or two quests. Otherwise it should just skip to the quest text (assuming the npc doesn't have the npcflag 1). So the linked image is wrong behavior.

These come from SMSG_QUEST_GIVER_QUEST_LIST_MESSAGE and are correct/blizzlike

this is not the same as gossip, which is dealt with separately

**Review by acidmanifesto — APPROVED** [review]

**Comment by Alastro-Kraken** [comment]
> these types of texts are not "gossips" per se either, as you cannot find any related gossip menu id's for them in sniffs
> this is not the same as gossip, which is dealt with separately from SMSG_GOSSIP_MESSAGE packets
Yeah I meant "gossip text" as "text that the npc shows in the gossip window". Sorry, that was confusing.

> But I do not understand what you mean by "So the linked image is wrong behavior" can you please explain?
In the image it shows the questgiver giving the questgiver greeting, but the player only has one quest he can take/turn in.
In that case it should not show the greeting but skip to the quest. The greeting is only shown when there are two or more quests/gossip options the player can see.

Example:
Right-clicking Foreman Oslow without any quest done (so only The Lost Tools is available):
WoWScrnShot_020422_231248


Right-clicking Foreman Oslow with [Ink Supplies](https://tbc.wowhead.com/quest=345/ink-supplies) in quest log (so The Lost Tools is available and Ink Supplies is in quest log):
WoWScrnShot_020422_231327

That's why I said the posted image is wrong - it shows the greeting while there's only one quest available and no gossip option.

**Comment by Kitzunu** [comment]
> That's why I said the posted image is wrong - it shows the greeting while there's only one quest available and no gossip option.

This is literally what this PR implements, quest_greetings will show a "quest greeting" even if there is only one quest available.
Because in some places the quest should not just "pop up", but there should be a gossip like view

cases like this doesn't have gossips in sniff, so this behavior is blizzlike

**Comment by Alastro-Kraken** [comment]
> This is literally what this PR implements, quest_greetings will show a "quest greeting" even if there is only one quest available.
> Because in some places the quest should not just "pop up", but there should be a gossip like view

When the NPC does not have npcflag gossip, and the player has more than one quest/gossip option with that NPC, then and only then it should show the quest greeting.

See [here](https://youtu.be/WO8Dm6JBrjM?t=206) - this NPC does not have npcflag gossip and has quest greeting, but it does not show that greeting - because that player can only accept one quest and there's no gossip options.

**Comment by Kitzunu** [comment]
quest greeting != gossips. Those are two different things.

If they were gossips we would not need to implement this, because gossips only show if there is > 1 option

**Comment by Alastro-Kraken** [comment]
I said gossip flag this time - that has everything to do with quest greetings.

Because if the NPC has gossip npcflag 1 - that means it has gossip and it's an actual gossip text, gossip menu and so on.
BUT if it does not have the gossip npcflag, talking to that NPC brings you to the text window/vendor window, UNLESS the player has access to more than one of those (let's say, a vendor with a quest). In that case, the quest greeting is shown.

**Comment by Kitzunu** [comment]
Maybe I am just slow today, and if so I apologize for that.

but you are essentially saying that quest greetings should work the same as gossips, to only show if there are > 1 option available?

**Comment by ZhengPeiRu21** [comment]
I think @Alastro-Kraken is correct. Here is a TrinityCore issue that has helpful related discussion and YouTube videos:
https://github.com/TrinityCore/TrinityCore/issues/19756

**Comment by Alastro-Kraken** [comment]
I'll go in more detail - it all depends on the npc flags.

If the NPC has the gossip npc flag - it's an actual gossip text, and it will show at any case. With one quest/gossip option, with two, with more or with less.

Quest greetings only affect NPCs that don't have the gossip npc flag. Those NPCs will skip any text and show you the quest directly, if the player can accept only one quest and there are no gossip options.

Take the human starting area NPC from the video I posted above: he does not have gossip npcflag, so he uses quest greeting. BUT he only has one quest for fresh level 1 humans. So, the player right-clicks, the npc shows the quest. No greeting.

So what happens if the NPC has TWO quests? Or if he has one quest and one gossip option? Or the player can turn in one quest, and accept another one from the NPC, at the same time? That's the only case where the quest greeting is shown.

**Comment by ZhengPeiRu21** [comment]
If this source is correct, it seems the change to skip the menu and not show the greetings text was made in TBC but could be enabled through a flag. Here is a thread with more details:
https://www.mmo-champion.com/threads/1467297-NPC-RP-text!-How-to-enable

**Comment by Kitzunu** [comment]
Okay, now I get it =)

**Comment by Kitzunu** [comment]
so basically we need to check if the creature has 
UNIT_NPC_FLAG_GOSSIP -> We always show quest greeting
if not -> We only show quest greeting while 2 or more options are available for the player.

Now my trick question. should it work the same for gameobjects?

**Comment by ZhengPeiRu21** [comment]
> so basically we need to check if the creature has UNIT_NPC_FLAG_GOSSIP -> We always show quest greeting if not -> We only show quest greeting while 2 or more options are available for the player.
> 
> Now my trick question. should it work the same for gameobjects?

The implementation of this table in Vmangos and TrinityCore only supports putting in creature entity ID. They do not have support for gameobjects to have quest_greeetings, is what I can find.

**Comment by Alastro-Kraken** [comment]
This is what I mean:

Uploading 2022-02-04 23-56-44.mp4…

See how the NPC does not show the greeting the first time, but shows it the second time?
Basically this change here https://github.com/azerothcore/azerothcore-wotlk/pull/10526/files#diff-5868e92ea22874f266f84d72837d190a784f74b38b6a04800dcdd85533813e15L152-R152 is wrong. Revert that and it should work perfectly :P

**Comment by Kitzunu** [comment]
https://trinitycore.atlassian.net/wiki/spaces/tc/pages/140771343/quest+greeting Type (0=Creature 1=GameObject)

**Comment by Kitzunu** [comment]
> See how the NPC does not show the greeting the first time, but shows it the second time?
Basically this change here https://github.com/azerothcore/azerothcore-wotlk/pull/10526/files#diff-5868e92ea22874f266f84d72837d190a784f74b38b6a04800dcdd85533813e15L152-R152 is wrong. Revert that and it should work perfectly :P

Yea I figured as much :D

**Comment by ZhengPeiRu21** [comment]
> https://trinitycore.atlassian.net/wiki/spaces/tc/pages/140771343/quest+greeting Type (0=Creature 1=GameObject)

Oh, sorry, I did not read carefully! Then I am mistaken. I do not know the answer.

**Comment by Kitzunu** [comment]
I removed that at first as I believed it to be that quest greetings always to show.

Thank you for the clarification!

**Comment by ZhengPeiRu21** [comment]
If the MMO Champion thread I linked is correct, I feel it is also proper behavior to see if client has that flag set, and if so, to always show gossip or quest greetings. But I think we may need a user to test that flag behavior on live servers to see how it is behave now. There are other references to ForceGossip, like this WoWHead comment: https://www.wowhead.com/npc=5512/kaita-deepforge#comments

**Comment by Alastro-Kraken** [comment]
> If the MMO Champion thread I linked is correct, I feel it is also proper behavior to see if client has that flag set, and if so, to always show gossip or quest greetings. But I think we may need a user to test that flag behavior on live servers to see how it is behave now. There are other references to ForceGossip, like this WoWHead comment: https://www.wowhead.com/npc=5512/kaita-deepforge#comments

There's no need to check anything, because the only NPCs that have quest greetings also don't have the gossip npc flag :P quest greetings exists only to give some dialogue to those NPCs that don't have gossip text normally.

About the forced gossip option thing, that's behavior that was implemented in one of the last WotLK patches (the same patch that moved the first questgivers in starting areas, I think?). The type flag CREATURE_TYPE_FLAG_FORCE_GOSSIP forces the creature to show gossip text even if there's only one option. There's nothing to fix on that front I think, it's sniffed data that should work automatically :D

**Comment by ZhengPeiRu21** [comment]
> > If the MMO Champion thread I linked is correct, I feel it is also proper behavior to see if client has that flag set, and if so, to always show gossip or quest greetings. But I think we may need a user to test that flag behavior on live servers to see how it is behave now. There are other references to ForceGossip, like this WoWHead comment: https://www.wowhead.com/npc=5512/kaita-deepforge#comments
> 
> There's no need to check anything server-side, because the only NPCs that have quest greetings also don't have the gossip npc flag :P quest greetings exists only to give some dialogue to those NPCs that don't have gossip text normally.
> 
> About the forced gossip option thing, that's behavior that was implemented in one of the last WotLK patches (the same patch that moved the first questgivers in starting areas, I think?). The type flag CREATURE_TYPE_FLAG_FORCE_GOSSIP forces the creature to show gossip text even if there's only one option. There's nothing to fix on that front I think, it's sniffed data that should work automatically :D

Thank you for explaining. Sorry to spend you time to teach me! Right now, from my work during recent PR, I feel that AzerothCore always is show the gossip even if there is only one option even if flag is not set, but that is a separate investigation. I will look more at it. On the other hand, it is not related to quest_greetings so I will discuss somewhere else! Thank you again.

**Comment by Kitzunu** [comment]
Okay, should be all good now!

**Comment by Alastro-Kraken** [comment]
> Thank you for explaining. Sorry to spend you time to teach me! Right now, from my work during recent PR, I feel that AzerothCore always is show the gossip even if there is only one option even if flag is not set, but that is a separate investigation. I will look more at it. On the other hand, it is not related to quest_greetings so I will discuss somewhere else! Thank you again.

No worries! I'm happy to help.

I remembered something - for a long while TrinityCore had that typeflag wrongly set to a whole lot of NPCs. This commit fixed it: https://github.com/TrinityCore/TrinityCore/commit/42bc12ddb6b2fd0b093f5189e60721ea989d7acc
If you do test this, make sure to clear your client's cache, because otherwise you won't see any change since this flag is cached by the client.

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by acidmanifesto** [comment]
@Kitzunu will u update the sql to include the populated tables from tc?

**Comment by Kitzunu** [comment]
No, for the simple reason that I cannot be bothered to recheck all entries and remove the gossips if we have them "fixed" that way

**Comment by ZhengPeiRu21** [comment]
Yes, quite a few of them are already this way "fixed". Once this branch is merged, I will check each one and populate the missing ones while fixing the ones using old work-around gossip menu, if you would like. (I will update my old PRs and create new ones for fixing issues - or I can do just one PR for further updates if prefer; I don't know which is easier to test, check source, etc.) Please let me know what is most helpful!

**Comment by Kitzunu** [comment]
Awesome, thanks!

I'd prefer one big PR for it

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 would you like to test this PR? :)

**Comment by ZhengPeiRu21** [comment]
> @ZhengPeiRu21 would you like to test this PR? :)

Yes, I will happy to test it! I may not have time to complete testing until Monday. I am busier with other things on weekend. If this is too long, please do not feel need to wait for me.

**Comment by Kitzunu** [comment]
I just need someone to test it (other than me) before it can be merged

**Comment by ZhengPeiRu21** [comment]
I have tested this code. It mostly seems to work. If the NPC has npcflag bit 1 set, they will show gossip_menu text, including showing placeholder "Greetings name" text even if they have a quest_greeting. I think this is correct behavior. NPC goes directly to quest text if only having one quest, and shows menu with multiple quest and quest_greeting text when bit is not set, so this is all looking correct.
Capture

However, the quest_greeting_locale does not seem to be working. It is still show the default text.
locale

This is with the DB looking like this:
quest_greeting
db

quest_greeting_locale
locale-db

I did clear client cache so I do not think this is a cache issue.

**Comment by Kitzunu** [comment]
I'll look into it tomorrow.

**Comment by Kitzunu** [comment]
try now

**Comment by ZhengPeiRu21** [comment]
Very sorry for the bad news, but it is still same behavior as before. It could be problem with Chinese client, but I only have English and Chinese to test, so I am not able to check. Other locale data does work with Chinese client, though.

**Comment by ZhengPeiRu21** [comment]
Thank you for spend so much time fix this issue. If we are still not able to find a working solution yet, maybe we can merge the code as is for now and open a separate issue for locale data not working in this table. Having the English text working would be a good improvement and would let to start adding the table data, even if there is still related issue to solve.

**Comment by Kitzunu** [comment]
Test now, for last hope lol

**Comment by ZhengPeiRu21** [comment]
> Test now, for last hope lol

Right now the code can't compile. There are a few errors in ObjectMgr:
```
azerothcore-wotlk\src\server\game\Globals\ObjectMgr.cpp(6190): error C2039: 'clear': is not a member of 'QuestGreetingLocale'
azerothcore-wotlk\src\server\game\Entities\Creature\CreatureData.h(349): note: see declaration of 'QuestGreetingLocale'
azerothcore-wotlk\src\server\game\Globals\ObjectMgr.cpp(6233): error C2676: binary '[': 'QuestGreetingLocale' does not define this operator or a conversion to a type acceptable to the predefined operator
azerothcore-wotlk\src\server\game\Globals\ObjectMgr.cpp(6233): error C2530: 'data': references must be initialized
```

I did clean my build directory before build, so I do not think problem is from my end. Sorry for the continued trouble!

**Comment by Kitzunu** [comment]
yep, is my bad. Will fix later tonight

**Comment by Kitzunu** [comment]
should be ok now

**Comment by ZhengPeiRu21** [comment]
Very sorry to still have a bad news, but it seems to still not be working! There is always chance that I am do some test wrong somehow, so maybe another tester with not enUS client can confirm.

**Comment by Kitzunu** [comment]
And you don't get any errors in worldserver?

**Comment by Kitzunu** [comment]
I dont have any client other than enUS so I can't debug it either. unfortunatly

**Comment by Kitzunu** [comment]
try again... otherwise im out of options... @acidmanifesto did you manage to debug it?

**Comment by ZhengPeiRu21** [comment]
> try again... otherwise im out of options... @acidmanifesto did you manage to debug it?

The new build does not run when there is data in the quest_greeting_locale table. It crashes with no error after `Loading Trainers...`. If the quest_greeting_locale table is empty, it works fine and the quest_greeting data appears as it should.

**Comment by temperrr** [comment]
image
image


Quest taking etc seems to work for me fine on enUS.

**Comment by Kitzunu** [comment]
Yea the enUS stuff works, just the locale things I need to figure out... But these days my time is slim and motivation low lol

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 can you test please

**Comment by Lynethris** [comment]
Work in French.
WoWScrnShot_022822_160614

**Comment by ZhengPeiRu21** [comment]
@Lynethris If it is not trouble, can you ensure that the text shown is for the locale used and not the default English text? I am just test again using zhTW client and it only is use the text from quest_greeting and not quest_greeting_locale:
Capture

**Comment by Kitzunu** [comment]
I really don't know how to fix it then... I use the same method that is used for any locale stuff

**Comment by ZhengPeiRu21** [comment]
> I really don't know how to fix it then... I use the same method that is used for any locale stuff

My thought is that I think we should merge this and can file a separate issue for locale data not working. There are many fix that we can do once this table is implemented, and having the locale data is a separate lower priority issue we can later fix. What do others think?

**Comment by Kitzunu** [comment]
let me try just one last thing

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 test now

**Comment by ZhengPeiRu21** [comment]
> @ZhengPeiRu21 test now

Sorry for a bad news, but it seem the build now is not working with this last change. Would you please take another look?

```
[ 22%] Building CXX object src/server/game/CMakeFiles/game.dir/cmake_pch.cxx.obj
cmake_pch.cxx
D:\Workshop\azerothcore-wotlk\src\server\game\Globals\ObjectMgr.h(1252): error C2039: 'find': is not a member of 'std::array<std::unordered_map<uint32,QuestGreetingLocale,std::hash<uint32>,std::equal_to<uint32>,std::allocator<std::pair<const uint32,QuestGreetingLocale>>>,2>'
D:\Workshop\azerothcore-wotlk\src\server\game\Globals\ObjectMgr.h(665): note: see declaration of 'std::array<std::unordered_map<uint32,QuestGreetingLocale,std::hash<uint32>,std::equal_to<uint32>,std::allocator<std::pair<const uint32,QuestGreetingLocale>>>,2>'
D:\Workshop\azerothcore-wotlk\src\server\game\Globals\ObjectMgr.h(1254): error C2039: 'second': is not a member of 'std::unordered_map<uint32,QuestGreetingLocale,std::hash<uint32>,std::equal_to<uint32>,std::allocator<std::pair<const uint32,QuestGreetingLocale>>>'
D:\Workshop\azerothcore-wotlk\src\server\game\Globals\ObjectMgr.h(665): note: see declaration of 'std::unordered_map<uint32,QuestGreetingLocale,std::hash<uint32>,std::equal_to<uint32>,std::allocator<std::pair<const uint32,QuestGreetingLocale>>>'
```

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 now

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -6112,6 +6113,132 @@ void ObjectMgr::LoadQuestAreaTriggers()
...
+        if (locale == LOCALE_enUS)
+            continue;
+
+        QuestGreetingLocale& data = _questGreetingLocaleStore[MAKE_PAIR32(type, id)];
+        AddLocaleString(fields[3].Get<std::string>(), locale, data.Greeting);
+
+        ++count;
+    } while (result->NextRow());
+
+    LOG_INFO("server.loading", ">> Loaded {} quest greeting locale strings in {} ms", (uint32)_questGreetingLocaleStore.count(), GetMSTimeDiffToNow(oldMSTime));
```
Comment (ZhengPeiRu21) [inline src/server/game/Globals/ObjectMgr.cpp:6239]:
```suggestion
    LOG_INFO("server.loading", ">> Loaded {} quest greeting locale strings in {} ms", (uint32)_questGreetingLocaleStore.size(), GetMSTimeDiffToNow(oldMSTime));
```
I have made this fix locally to allow build to complete. I will let you know if locale data now is working!

**Comment by ZhengPeiRu21** [comment]
It is still not working. However, there is now useful information in the log that may help solve the problem:

```
Table `quest_greeting_locale`: creature template entry 265 does not exist.
>> Loaded 0 quest greeting locale strings in 0 ms
```

**Comment by Kitzunu** [comment]
And that is fucked as that creature does exist, and does pass the exact same check in the quest_greetings load 🤷

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 try now

**Comment by ZhengPeiRu21** [comment]
It is now loading correctly and I am seeing this in the startup logs:
```
Loading Quests Greetings...
>> Loaded 1 quest_greeting in 0 ms
>> Loaded 1 quest greeting locale strings in 1 ms
```

However, the game is still only show the English text. (I did clear my Cache to confirm it is not cache issue.)

**Comment by Kitzunu** [comment]
@ZhengPeiRu21 you dont see locale loading in startup logs?

File: src/server/game/World/World.cpp [hunk 2]
```diff
@@ -1732,6 +1732,11 @@ void World::SetInitialWorldSettings()
     LOG_INFO("server.loading", "Loading Quests Starters and Enders...");
     sObjectMgr->LoadQuestStartersAndEnders();                    // must be after quest load
 
+    LOG_INFO("server.loading", "Loading Quest Greetings...");
+    sObjectMgr->LoadQuestGreetings();                               // must be loaded after creature_template, gameobject_template tables
+    LOG_INFO("server.loading", "Loading Quest Greeting Locales...")
```
Comment (Kitzunu) [inline src/server/game/World/World.cpp:1737]:
```suggestion
    LOG_INFO("server.loading", "Loading Quest Greeting Locales...");
```

**Comment by ZhengPeiRu21** [comment]
> @ZhengPeiRu21 you dont see locale loading in startup logs?

Sorry, I test in between when you said test and when you changed the logging message. I will test again with latest version.

**Comment by Kitzunu** [comment]
Im out of ideas now. For real

**Comment by Kitzunu** [comment]
We merge this and fix locales later

**Comment by acidmanifesto** [comment]
code factor is being weird. it says fail but when you click on it, it says no issues found.
image

**Pull request merged**